hibernate-core/src/main/java/org/hibernate/action/internal/EntityInsertAction.java

```diff
@@ -150,6 +153,168 @@ public void execute() throws HibernateException {
 
 		handleNaturalIdPostSaveNotifications( id );
 
+		EntityMetamodel entityMetamodel = getPersister().getEntityMetamodel();
+
+		if (entityMetamodel.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading()) {
+			PersistentAttributeInterceptable interceptable = ((PersistentAttributeInterceptable) instance );
+			PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
+			final BytecodeEnhancementMetadata enhancementMetadata = entityMetamodel.getBytecodeEnhancementMetadata();
+			final LazyAttributesMetadata lazyAttributesMetadata = enhancementMetadata.getLazyAttributesMetadata();
+			if (interceptor == null) {
+				Type[] types = persister.getPropertyTypes();
+				final String[] propertyNames = persister.getPropertyNames();
+				final CascadeStyle[] cascadeStyles = persister.getPropertyCascadeStyles();
+				boolean needsInterceptor = false;
+				Set<String> initializedLazyFields = new HashSet<>();
+				for ( int i = 0; i < types.length; i++) {
+					if(!types[i].isAssociationType()) {
+						continue;
+					}
+					final String propertyName = propertyNames[i];
+					CascadeStyle cascadeStyle = cascadeStyles[i];
+					if(cascadeStyle != CascadeStyles.NONE) {
+						initializedLazyFields.add(propertyName);
+						continue;
+					}
+					Object propertyValue = persister.getEntityTuplizer().getGetter(i).get(instance);
+					if(propertyValue == null) {
+						if(types[i].isCollectionType()) {
+							needsInterceptor = true;
+							continue;
+						}
+					}
+					else {
+						if(propertyValue instanceof PersistentAttributeInterceptable) {
+							PersistentAttributeInterceptable interceptableProperty = ((PersistentAttributeInterceptable) propertyValue );
+							PersistentAttributeInterceptor propertyInterceptor = interceptableProperty.$$_hibernate_getInterceptor();
+							if(propertyInterceptor == null) {
+								needsInterceptor = true;
+								continue;
+							}
+							else {
+								if(lazyAttributesMetadata.isLazyAttribute(propertyName)) {
+									initializedLazyFields.add(propertyName);
+								}
+							}
+						}
+						if(types[i].isCollectionType()) {
+							if(lazyAttributesMetadata.isLazyAttribute(propertyName)) {
+								initializedLazyFields.add(propertyName);
+							}
+							continue;
+						}
+					}
+				}
+				if (needsInterceptor) {
+					persister.getBytecodeEnhancementMetadata().injectEnhancedEntityAsProxyInterceptor(instance, getEntityKey(), session);
+					interceptor = interceptable.$$_hibernate_getInterceptor();
+					for (String propertyName : initializedLazyFields) {
```

Comment:
True, but most of this code does not execute if the property is not an association type:

`if( ! types[i].isAssociationType()) continue;`

Still, I definitely agree with you, a few iterations can be saved, and I thought about this as I was writing this code.  However, since I am not on the ORM team, I have chosen to avoid creating new class member variables, and changing the signature of existing methods, and just tried to work within the framework.  I wanted the ORM team to decide where and how to optimize, especially since ORM 6.0 is a major release.  I was willing to do 95% of the work to get this up and running in ORM 5.x, but the remaining 5% to fully optimize might double our investment, and them might be thrown out when we see the changes in ORM 6.0?  Ideally, we should try to interrupt the ORM team to look at this while 6.0 is unreleased so they can introduce any needed breaking changes at the major release, rather than at a point release.